pandas/tests/indexing/test_datetime.py

```diff
@@ -356,3 +356,19 @@ def test_loc_label_slicing(self):
         expected = ser.iloc[:-1]
 
         tm.assert_series_equal(result, expected)
+
+    def test_apply_datetime_tz_issue(self):
+        # GH 29052
+
+        df = DataFrame(
+            data=[0, 1, 2],
+            index=[
+                pd.Timestamp("2019-03-15 12:34:31.909000+0000", tz="UTC"),
+                pd.Timestamp("2019-03-15 12:34:34.359000+0000", tz="UTC"),
+                pd.Timestamp("2019-03-15 12:34:34.660000+0000", tz="UTC"),
+            ],
+        )
+        df2 = df.apply(lambda x: x.name, axis=1)
+
+        assert df[0].keys().dtype == df2.dtypes
+        assert (df[0].keys().values == df2.values).all()
```

Comment:
construct the expected DataFrame explicitly and do `tm.assert_frame_equals(result,expected)`

Reply:
> @GrigoriosGiann Thanks for the PR.

Thanks for the feedback! 😄
Let me know if there is anything else I should adjust. 🙂